controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java

```diff
@@ -862,9 +862,30 @@ private HttpResponse deploy(String tenantName, String applicationName, String in
             applicationPackage = Optional.of(controller.applications().getApplicationPackage(controller.applications().require(applicationId), applicationVersion.get()));
         }
 
+
+        boolean deployDirectly = deployOptions.field("deployDirectly").asBool();
+        Optional<Version> vespaVersion = optional("vespaVersion", deployOptions).map(Version::new);
+
+        if(deployDirectly && !applicationPackage.isPresent() && !applicationVersion.isPresent() && !vespaVersion.isPresent()) {
+            // Redeploy the existing deployment with the same versions.
+            Optional<Deployment> deployment = controller.applications().get(applicationId)
+                    .map(Application::deployments)
+                    .flatMap(deployments -> Optional.ofNullable(deployments.get(zone)));
+
+            if(!deployment.isPresent())
+                throw new IllegalArgumentException("Can't redeploy application, no deployment currently exist");
+
+            applicationVersion = Optional.ofNullable(deployment.get().applicationVersion());
```

Comment:
`.applicationVersion()` never returns `null`. What you want to check for is whether it's `ApplicationVersion.unknown`.